File: src/tier1/library/typescript/src/secret.ts
/**
 * secret.ts — k1s0 tier1 Library TypeScript 実装: Secret Management L3 facade interface
 * 05_鍵管理適合仕様.md §v1 key_class セット（5 class）および §SecretStore 抽象 に準拠する。
 * Rust core/secret.rs の SecretStore trait と 4 言語等価強度を保つ。
 * OpenBao / AWS Secrets Manager 等のバックエンドを実装で切り替えられる interface を宣言する。
 * 公開 API シグネチャに生 key bytes / 生シークレット値を露出しない。
 */

// KeyClass を keyHandle.ts から参照する（4 言語等価のため同一 enum を使う）
import type { KeyClass } from "./keyHandle.js";

/**
 * SecretMetadata はシークレットのメタデータを宣言する型。
 * 生のシークレット値は含まない（SecretStore.withSecret の callback パターンで別管理する）。
 * Rust core/secret.rs の SecretMetadata struct と 1:1 対応する。
 */
// SecretMetadata 型定義: シークレットのメタデータのみを保持する
export interface SecretMetadata {
  // secretId: シークレットの識別子（UUID v7 形式）
  readonly secretId: string;
  // version: シークレットのバージョン番号（ローテーション追跡用）
  readonly version: number;
  // keyClass: このシークレットを保護している鍵のクラス（05_鍵管理適合仕様 §v1 5 class）
  readonly keyClass: KeyClass;
  // tenantId: シークレットが属するテナントの識別子
  readonly tenantId: string;
  // isActive: シークレットが有効かどうか（revoke / rotate 後 false になる）
  readonly isActive: boolean;
}

/**
 * SecretRotationPolicy はシークレットローテーションポリシーを宣言する型。
 * Rust core/secret.rs の SecretRotationPolicy enum と 4 言語等価強度を保つ。
 */
// SecretRotationPolicy 型定義: ローテーション方式を宣言する
export type SecretRotationPolicy =
  // manual: 手動ローテーション（自動ローテーションなし）
  | { readonly kind: "manual" }
  // onDemand: 要求時ローテーション（呼び出し元が rotate を呼ぶ）
  | { readonly kind: "onDemand" }
  // scheduled: スケジュールローテーション（間隔秒数で指定）
  | { readonly kind: "scheduled"; readonly intervalSeconds: number };

/**
 * SecretStore は Secret Management の L3 抽象 interface。
 * OpenBao / AWS Secrets Manager 等の OSS / サービスを実装で切り替えられる。
 * Rust core/secret.rs の SecretStore trait と 4 言語等価強度を保つ。
 * 公開 API に生シークレット値を返さない（withSecret の callback パターンを使う）。
 */
// SecretStore インターフェース定義
export interface SecretStore {
  /**
   * getMetadata はシークレットのメタデータのみを返す（値は返さない）。
   * Rust の get_metadata(&self, secret_id, tenant_id) -> Result<Option<SecretMetadata>> に対応する。
   */
  // getMetadata メソッド: シークレットのメタデータのみを返す（値は返さない）
  getMetadata(
    // secretId: シークレットの識別子（UUID v7 形式）
    secretId: string,
    // tenantId: テナント識別子（RLS 境界を越えないことを保証する）
    tenantId: string
  ): Promise<SecretMetadata | null>;

  /**
   * withSecret はシークレット値を callback に渡して処理させる。
   * callback 外にシークレット値が漏れない設計（値は返さない）。
   * Rust の with_secret(&self, secret_id, tenant_id, f: FnOnce(&SecretValue) -> T) -> Result<T> に対応する。
   */
  // withSecret メソッド: シークレット値を callback に渡す（値は返さない）
  withSecret<T>(
    // secretId: シークレットの識別子（UUID v7 形式）
    secretId: string,
    // tenantId: テナント識別子（RLS 境界を越えないことを保証する）
    tenantId: string,
    // callback: シークレットのバイト列を受け取り T を返す処理（バイト列を漏洩させない）
    callback: (secretBytes: Readonly<Uint8Array>) => T
  ): Promise<T>;

  /**
   * rotate はシークレットを新しいバージョンにローテーションする。
   * 返した SecretMetadata に新しい version が含まれる。
   * Rust の rotate(&self, secret_id, tenant_id) -> Result<SecretMetadata> に対応する。
   */
  // rotate メソッド: シークレットを新しいバージョンにローテーションする
  rotate(
    // secretId: シークレットの識別子（UUID v7 形式）
    secretId: string,
    // tenantId: テナント識別子（RLS 境界を越えないことを保証する）
    tenantId: string
  ): Promise<SecretMetadata>;

  /**
   * revoke はシークレットを無効化する（isActive=false にする）。
   * revoke 後の withSecret 呼び出しは例外を投げる。
   * Rust の revoke(&self, secret_id, tenant_id) -> Result<()> に対応する。
   */
  // revoke メソッド: シークレットを無効化する
  revoke(
    // secretId: シークレットの識別子（UUID v7 形式）
    secretId: string,
    // tenantId: テナント識別子（RLS 境界を越えないことを保証する）
    tenantId: string
  ): Promise<void>;
}

File: src/tier1/library/rust/src/secret.rs
// secret.rs — k1s0 tier1 Library: シークレット管理 L1+ facade trait
// OpenBao 等の OSS 型を公開 API に露出しない（L1+ ラップ規約）。
// 生シークレット値を公開 API に露出禁止（SecretHandle による opaque ラップを必須とする）。
// 全 trait は Send + Sync を要求する（スレッド安全性の強制）。

// async_trait: async fn in trait を stable で使用するためのマクロ
use async_trait::async_trait;

// SecretStore は OpenBao を L1+ ラップするシークレット管理 facade trait。
// 公開 API シグネチャに OSS 型（vaultrs::client::VaultClient 等）を一切含まない。
// 取得したシークレット値は SecretHandle にラップして返す（生 bytes 露出禁止）。
#[async_trait]
pub trait SecretStore: Send + Sync {
    // get_secret はシークレット名で値を取得して SecretHandle にラップして返す。
    // name はシークレット識別子（例: "db/password" / "jwt/signing_key"）。
    // 戻り値は SecretHandle（生シークレット bytes は公開 API に露出しない）。
    async fn get_secret(&self, name: &str) -> crate::Result<SecretHandle>;

    // put_secret はシークレットを設定する（値は SecretHandle で渡す）。
    // name はシークレット識別子（例: "db/password"）。
    // secret は SecretHandle（生 bytes を直接受け取らない設計で規律を強制する）。
    async fn put_secret(&self, name: &str, secret: SecretHandle) -> crate::Result<()>;
}

// SecretHandle は生シークレット値を保持しない opaque ハンドル型。
// pub(crate) のみで内部アクセスを許可し、公開 API への生 bytes 露出を禁止する。
// drop 時のゼロクリアは実装の KeyMaterial と同様のパターンで対応すること。
pub struct SecretHandle(Vec<u8>);

// SecretHandle のコンストラクタと内部アクセサ（pub(crate) で公開 API 露出禁止）
impl SecretHandle {
    // from_raw は内部（opaque）値を生成する — 実装内部のみ使用可。
    // pub(crate) 限定のため、外部クレートから生 bytes を直接 SecretHandle に変換できない。
    pub(crate) fn from_raw(raw: Vec<u8>) -> Self {
        // Vec<u8> をそのまま保持する（opaque ラップ）
        Self(raw)
    }

    // as_bytes は内部値への参照を返す — pub(crate) のみ許可（公開 API への露出禁止）。
    // 外部クレートからは生 bytes にアクセスできない（ゼロ知識原則の強制）。
    pub(crate) fn as_bytes(&self) -> &[u8] {
        // 内部バイト列への参照を返す
        &self.0
    }
}
